rust/xaynet-sdk/src/state_machine/phase.rs

```diff
@@ -0,0 +1,223 @@
+                self.io.notify_new_round();
+                TransitionOutcome::Complete(
+                    Phase::<NewRound>::new(State::new(self.state.shared, NewRound), self.io).into(),
+                )
+            }
+            RoundFreshness::Fresh => {
+                debug!("round is still fresh, continuing from where we left off");
+                <Self as Step>::step(self).await
+            }
+        }
+    }
+
+    async fn check_round_freshness(&mut self) -> RoundFreshness {
+        match self.io.get_round_params().await {
+            Err(e) => {
+                warn!("failed to fetch round parameters {:?}", e);
+                RoundFreshness::Unknown
+            }
+            Ok(params) => {
+                if params == self.state.shared.round_params {
+                    debug!("round parameters didn't change");
+                    RoundFreshness::Fresh
+                } else {
+                    info!("fetched fresh round parameters");
+                    self.state.shared.round_params = params;
+                    RoundFreshness::Outdated
+                }
+            }
+        }
+    }
+}
+
+impl<P> Phase<P> {
+    pub(in crate::state_machine) fn new(
+        state: State<P>,
+        io: Box<dyn IO<Model = Box<dyn AsRef<Model> + Send>>>,
+    ) -> Self {
+        Self { state, io }
+    }
+}
+
+impl<P> Phase<P> {
+    pub(super) fn into_awaiting(self) -> Phase<Awaiting> {
+        Phase::<Awaiting>::new(State::new(self.state.shared, Awaiting), self.io)
+    }
+
+    pub(super) async fn send_message(
+        &mut self,
+        encoder: MessageEncoder,
+    ) -> Result<(), SendMessageError> {
+        for part in encoder {
+            let data = self.state.shared.round_params.pk.encrypt(part.as_slice());
+            self.io.send_message(data).await.map_err(|e| {
+                error!("failed to send message: {:?}", e);
+                SendMessageError
+            })?
+        }
+        Ok(())
+    }
+
+    pub(super) fn message_encoder(&self, payload: Payload) -> MessageEncoder {
+        MessageEncoder::new(
+            self.state.shared.keys.clone(),
+            payload,
+            self.state.shared.round_params.pk,
+            self.state
+                .shared
+                .message_size
+                .max_payload_size()
+                .unwrap_or(0),
+        )
+        // the encoder rejects Chunk payload, but in the state
+        // machine, we never manually create such payloads so
+        // unwrapping is fine
+        .unwrap()
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("failed to send a PET message")]
+pub struct SendMessageError;
+
+pub enum RoundFreshness {
+    Outdated,
+    Unknown,
+    Fresh,
+}
+
+/// A serializable representation of a phase state.
+#[derive(Serialize, Deserialize, From)]
+pub enum SerializableState {
+    NewRound(State<NewRound>),
+    Awaiting(State<Awaiting>),
+    Sum(State<Sum>),
+    // FIXME: this should be boxed...
```

Comment:
is this FIXME still valid?

Reply:
Yes. I could easily be done in a follow-up PR. Perhaps something that @wilk10 could tackle to dive in the client code?